core/chains/evm/logpoller/log_poller_internal_test.go

```diff
@@ -0,0 +1,316 @@
+	require.False(t, ok, "'Emitter Log 1 Filter' should have been removed by UnregisterFilter()")
+	require.Len(t, lp.filters, 2)
+	validateFiltersTable(t, lp, orm)
+
+	err = lp.UnregisterFilter("Emitter Log 1 + 2 dupe")
+	require.NoError(t, err)
+	err = lp.UnregisterFilter("Emitter Log 1")
+	require.NoError(t, err)
+	assert.Len(t, lp.filters, 0)
+	filters, err := lp.orm.LoadFilters()
+	require.NoError(t, err)
+	assert.Len(t, filters, 0)
+
+	// Make sure cache was invalidated
+	assert.Len(t, lp.Filter(nil, nil, nil).Addresses, 1)
+	assert.Equal(t, lp.Filter(nil, nil, nil).Addresses[0], common.HexToAddress("0x0000000000000000000000000000000000000000"))
+	assert.Len(t, lp.Filter(nil, nil, nil).Topics, 1)
+	assert.Len(t, lp.Filter(nil, nil, nil).Topics[0], 0)
+}
+
+func assertForeignConstraintError(t *testing.T, observedLog observer.LoggedEntry,
+	table string, constraint string) {
+
+	assert.Equal(t, "SQL ERROR", observedLog.Entry.Message)
+
+	field := observedLog.Context[0]
+	require.Equal(t, zapcore.ErrorType, field.Type)
+	err, ok := field.Interface.(error)
+	var pgErr *pgconn.PgError
+	require.True(t, errors.As(err, &pgErr))
+	require.True(t, ok)
+	assert.Equal(t, "23503", pgErr.SQLState()) // foreign key constraint violation code
+	assert.Equal(t, table, pgErr.TableName)
+	assert.Equal(t, constraint, pgErr.ConstraintName)
+}
+
+func TestLogPoller_DBErrorHandling(t *testing.T) {
+	t.Parallel()
+	lggr, observedLogs := logger.TestLoggerObserved(t, zapcore.WarnLevel)
+	chainID1 := testutils.NewRandomEVMChainID()
+	chainID2 := testutils.NewRandomEVMChainID()
+	db := pgtest.NewSqlxDB(t)
+	o := NewORM(chainID1, db, lggr, pgtest.NewQConfig(true))
+
+	owner := testutils.MustNewSimTransactor(t)
+	ethDB := rawdb.NewMemoryDatabase()
+	ec := backends.NewSimulatedBackendWithDatabase(ethDB, map[common.Address]core.GenesisAccount{
+		owner.From: {
+			Balance: big.NewInt(0).Mul(big.NewInt(10), big.NewInt(1e18)),
+		},
+	}, 10e6)
+	_, _, emitter, err := log_emitter.DeployLogEmitter(owner, ec)
+	require.NoError(t, err)
+	_, err = emitter.EmitLog1(owner, []*big.Int{big.NewInt(9)})
+	require.NoError(t, err)
+	_, err = emitter.EmitLog1(owner, []*big.Int{big.NewInt(7)})
+	require.NoError(t, err)
+	ec.Commit()
+	ec.Commit()
+	ec.Commit()
+
+	lp := NewLogPoller(o, client.NewSimulatedBackendClient(t, ec, chainID2), lggr, 1*time.Hour, 2, 3, 2, 1000)
+	ctx, cancelReplay := context.WithCancel(testutils.Context(t))
+	lp.ctx, lp.cancel = context.WithCancel(testutils.Context(t))
+	defer cancelReplay()
+	defer lp.cancel()
+
+	err = lp.Replay(ctx, 5) // block number too high
+	require.ErrorContains(t, err, "Invalid replay block number")
+
+	// Force a db error while loading the filters (tx aborted, already rolled back)
+	require.Error(t, utils.JustError(db.Exec(`invalid query`)))
+	go func() {
+		err = lp.Replay(ctx, 2)
+		assert.Error(t, err, ErrReplayAbortedByClient)
+	}()
+
+	time.Sleep(100 * time.Millisecond)
+	go lp.run()
+	require.Eventually(t, func() bool {
```

Comment:
usually best to avoid sleep/eventually/lp.run if we can - is there no way to make it synchronous @reductionista?

Reply:
If we want to have error handling in `lp.run()`, then the only way to get coverage of those error conditions is to call it.

I do think that `lp.run()` has far too much complexity in it; one thing I've thought about is we should strip that down to be mostly a series of calls to other functions, with very minimal logic there.  I don't think we could remove everything though.  This is a problem I've seen in lots of services--I think it's why we don't usually have test for error conditions in run loops.  But SonarCube now complains unless you have at least 90% coverage, which was impossible in this case without covering the error conditions.

Reply:
I think `require.Eventually()` is the optimal solution to the problem.  It allows us to test run loops without  introducing any extra sleep delays as long as the test is passing, but does it in a nice clean synchronized way with no possibility for race conditions.  I think the only thing that could do as well would be some kind of channel communication between the test and the inside of the runloop, but that would make the production code unnecessarily complex.

The call to `time.Sleep(100 * time.Millisecond)` is more ugly I'd say.  There might be a way to improve that--I tried to think of some, but couldn't find anything simple.  I can give it some more thought.